### DISCUSSION TITLE
Create a Policy for Ec2 instance which doesn't have name

### DISCUSSION THREAD

**Discussion by agaur123** [cloud-custodian/cloud-custodian#7330] [General]
hi, I have written policy for ec2 instance which doesn't have a name and while dry run run mode, it's working fine, but I am confused that how we have to handle the auto scaling groups.
Code for Ec2
- name: ec2-non-name-compliance
  resource: ec2
  description: |
    Schedule a resource that does not meet name compliance policies.
  mode:
      type: periodic
      schedule: "rate(1 Day)"
  filters:
   - "tag:Name": absent
  actions:
    - type: mark-for-op
      tag: custodian_status
      op: terminate
      days: 4

**Comment by kapilt** [comment]
its best to have a separate policy for autoscaling groups, instances in an autoscaling group have a tag placed on them by the autoscaling group, then you can exempt the asg instances from your standard instance policies by checking for the asg tag (prefixed 'aws') so they can be handled by your asg policies.

**Comment by agaur123** [comment]
So what i have understood from your explanation is 
policies:
- name: ec2-non1-name-compliance
  resource: ec2
  description: |
    Schedule a resource that does not meet name compliance policies.
  mode:
      type: periodic
      schedule: "rate(1 Day)"
  filters:
   - "tag:Name": absent
  actions:
    - type: mark-for-op
      tag: custodian_status
      op: terminate
      days: 4

- name: asg-non-name-compliance
  resource: ec2
  description: |
    Schedule a resource that does not meet name compliance policies.
  mode:
      type: periodic
      schedule: "rate(1 Day)"
  filters:
   - "tag:Name": absent
   - "tag:aws:autoscaling:groupName": Present
  actions:
    - type: mark-for-op
      tag: custodian_status
      op: terminate
      days: 4

**Comment by agaur123** [comment]
One more thing that if any instance didn't have any name and it's created through ASG, than the policy should terminate the ASGroup, not the Instance.

**Comment by thisisshi** [comment]
@agaur123 you'll need a separate asg policy to handle that, you can suspend processes on the asg level to prevent it from creating new instances see:

```shell
$ custodian schema asg.actions.suspend
```

**Comment by agaur123** [comment]
@thisisshi the explanation for the command says that your ASG should have Suspend Tag, but what if we don't have that while creating the ASG, than we have to handle that ?
`
:example:

.. code-block:: yaml

        policies:
          - name: asg-suspend-processes
            resource: asg
            filters:
              - "tag:SuspendTag": present
            actions:
              - type: suspend`

**Comment by thisisshi** [comment]
@agaur123 no, that's just an arbitrary filter

your policy would look something like:

```yaml
policies:
  - name: suspend-asg-without-name
    resource: asg
    filters:
      - "tag:Name": absent
    actions:
      - type: suspend
```

**Comment by agaur123** [comment]
@thisisshi See let me Tell you the whole scenario, I want to make policy to delete the Ec2 instance which doesn't have the name(This can be easy), then if some instance is not having the name and it's created through ASG, then in that case, we have to delete the ASG, so that automatically the Instance will delete. The thing is we don't have to make a policy for ASG without the name, we have to make policy for instance which is made through ASG and then suspend that ASG.

**Comment by thisisshi** [comment]
I understand, what I'm saying is you need to have a policy to both terminate ec2s that aren't connected to an asg and a policy to suspend the asg autoscaling process and then terminate the no-name ec2's. 

The suspend action will prevent the asg from creating new ec2 instances and it should only have to check if the Name tag is absent as tags are propagated from the asg to the ec2 instance. Once the asgs with no name are suspended then you can safely terminate the rest of the no-name ec2 instances.

**Comment by agaur123** [comment]
SO from your Explanation i have written the policy
` policies:
- name: ec2-non-asg-compliance
  resource: ec2
  description: |
   Resource that does not have name tag
  mode:
      type: periodic
      schedule: "rate(1 Day)"
  filters:
   - "tag:aws:autoscaling:groupName": absent
  actions:
    - type: mark-for-op
      tag: custodian_status
      op: terminate
      days: 4

- name: suspend-asg-without-name
  resource: asg
  description: |
   Resource that does not have name tag
  filters:
      - "tag:Name": absent
  actions:
      - type: suspend

- name: ec2-non-name-compliance
  resource: ec2
  description: |
   Resource that does not have name tag
  mode:
      type: periodic
      schedule: "rate(1 Day)"
  filters:
   - "tag:Name": absent
  actions:
    - type: mark-for-op
      tag: custodian_status
      op: terminate
      days: 4
`

**Comment by agaur123** [comment]
tell me if i need to make any change in this

**Comment by agaur123** [comment]
and one thing this way we are terminating the name Ec2 Instances, and we dont want to do that. Just want for No-name Ec2 Instances.
